suite/tests/api/ir_regdeps.c

```diff
@@ -0,0 +1,325 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "configure.h"
+#include "dr_api.h"
+#include "tools.h"
+
+#define ASSERT(x)                                                                 \
+    ((void)((!(x)) ? (dr_fprintf(STDERR, "ASSERT FAILURE: %s:%d: %s\n", __FILE__, \
+                                 __LINE__, #x),                                   \
+                      dr_abort(), 0)                                              \
+                   : 0))
+
+#define ALIGN_BYTES 4
+
+static void
+test_instr_encode_decode_synthetic(void *dc, instr_t *instr)
+{
+    /* Encoded synthetic ISA instructions require 4 byte alignment.
+     * The largest synthetic encoded instruction has 16 bytes.
+     */
+    byte ALIGN_VAR(ALIGN_BYTES) bytes[16];
+
+    /* Convert a real ISA instruction to a synthetic ISA (DR_ISA_REGDEPS) instruction.
+     */
+    instr_t *instr_synthetic_converted = instr_create(dc);
+    instr_convert_to_isa_regdeps(dc, instr, instr_synthetic_converted);
+
+    /* Encode the synthetic instruction.
+     */
+    byte *next_pc_encode = instr_encode(dc, instr_synthetic_converted, bytes);
+    ASSERT(next_pc_encode != NULL);
+
+    /* Create an instruction where we can decode the previously encoded synthetic
+     * instruction.
+     */
+    instr_t *instr_synthetic_decoded = instr_create(dc);
+
+    dr_isa_mode_t old_isa_mode;
+    /* DR uses dcontext_t ISA mode to decode instructions.
+     * Since we are decoding synthetic instructions, we set it to DR_ISA_REGDEPS.
+     */
+    dr_set_isa_mode(dc, DR_ISA_REGDEPS, &old_isa_mode);
+    /* Decode the encoded synthetic instruction bytes into instr_synthetic.
+     */
+    byte *next_pc_decode = decode(dc, bytes, instr_synthetic_decoded);
+    dr_set_isa_mode(dc, old_isa_mode, NULL);
+    ASSERT(next_pc_decode != NULL);
+    ASSERT(next_pc_encode == next_pc_decode);
+    /* Check for overflow.
+     */
+    ASSERT((next_pc_encode - bytes) <= sizeof(bytes));
+    ASSERT((next_pc_decode - bytes) <= sizeof(bytes));
+    /* Check that the two synthetic instructions are the same.
+     */
+    ASSERT(instr_same(instr_synthetic_converted, instr_synthetic_decoded));
+
+    instr_destroy(dc, instr);
+    instr_destroy(dc, instr_synthetic_converted);
+    instr_destroy(dc, instr_synthetic_decoded);
+}
+
+#ifdef X86_64
+static void
+test_instr_create_encode_decode_synthetic_x86_64(void *dc)
+{
+    byte buf[128];
+    instr_t *instr;
+
+    instr = INSTR_CREATE_push(dc, opnd_create_reg(SEG_FS));
+    /* Instructions generated by INSTR_CREATE_ or XINST_CREATE_ are not fully decoded.
+     * For example, their categories are not set.  So we encode and decode them to obtain
+     * a fully decoded instruction before testing our DR_ISA_REGDEPS synthetic encoding
+     * and decoding.  We do so for all instructions of all architectures.
```

Comment:
Just wanted to point this out.
We do the same in suite/tests/api/drdecode_x86.c when checking for categories (https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/api/drdecode_x86.c#L166).
Is there a better way or is this ok?

Reply:
This is a wart from the original category feature which IMHO leaves that feature incomplete but that is the current status so yes.

Reply:
Got it.